### PR TITLE
Move interners from trait to generic structs

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This rewrites the interner mechanisms be defined in terms of two generic structs (`Interner` and `Interned`) that formalise the idea that the interned values are owned versions of a given reference type, and move the logically separate intern-related actions (get a key from the value, and get the value from the key) into separate functions (`get` and `insert`, respectively).

This has a few advantages:

1. we now have both `insert` and `insert_owned` methods, which was awkward to add within the trait-based structure.  This allows a more efficient path when the owned variant has already necessarily been constructed.

2. additionally, the standard `insert` path now takes only a reference type.  For large circuits, most intern lookups will, in general, find a pre-existing key, so in situations where the interned value is sufficiently small that it can be within a static allocation (which is the case for almost all `qargs` and `cargs`), it's more efficient not to construct the owned type on the heap.

3. the type of the values retrieved from an interner are no longer indirected through the owned type that's stored.  For example, where `IndexedInterner<Vec<Qubit>>` gave out `&Vec<Qubit>`s as its lookups, `Interner<[Qubit]>` returns the more standard `&[Qubit]`, which is only singly indirect rather than double.

The following replacements are made:

1. The `IndexedInterner` struct from before is now just called `Interner` (and internally, it uses an `IndexSet` rather than manually tracking the indices).  Its generic type is related to the references it returns, rather than the owned value it stores, so `IndexedInterner<Vec<Qubit>>` becomes `Interner<[Qubit]>`.

2. The `Interner` trait is now gone.  Everything is just accessed as concrete methods on the `Interner` struct.

3. `<&IndexedInterner as Interner>::intern` (lookup of the value from an interner key) is now called `Interner::get`.

4. `<&mut IndexedInterner as Interner>::intern` (conversion of an owned value to an interner key) is now called `Interner::insert_owned`.

5. A new method, `Interner::insert`, can now be used when one need not have an owned allocation of the storage type; the correct value will be allocated if required (which is expected to be less frequent).

6. The intern key is no longer called `interner::Index`, but `Interned<T>`, where the generic parameter `T` matches the generic of the `Interner<T>` that gave out the key.

### Details and comments

This is sufficiently generic now that it can probably be used directly for the purposes of #12917, though I didn't make it generic over the key width.  It probably wouldn't be too hard to do that too, if we want it very much.

I'm intending to follow this PR with one that makes the key corresponding to the "empty" state constructible without any hash lookup, since we end up needing that quite a lot (think the cargs key when pushing a standard gate). (*edit*: now #13035.)

### CODE REVIEW

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]: Perhaps this would be a good time for us to also rename `{qargs,cargs}_cache` to `{qargs,cargs}_interner` to match the naming in `CircuitData`?

(I meant to do that originally, but it got away from me before we merged `DAGCircuit`.)

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:1678]: So in this case, we can use `collect` here which will collect into a native array? This seems very cool, but I think I'm not understanding how this works. Can you explain?

File: crates/circuit/src/interner.rs
Comment (kevinhartman) [inline crates/circuit/src/interner.rs:33]: Can you explain (mostly just to help me understand) why you're using `*const T` as the type parameter instead of just `T`? It looks like `PhantomData`'s type parameter just has to be `?Sized`.

**Review by kevinhartman — COMMENTED** [review]
I love this.

Very cool Rust going on here that gives the interner a much nicer interface. Super curious to hear the answers to some of the questions I've left in line.

Thanks for doing this, @jakelishman!

File: crates/circuit/src/dag_circuit.rs
Comment (jakelishman) [reply crates/circuit/src/dag_circuit.rs:1678]: This might be more straightforward than you had in mind - `<[T] as ToOwned>::Owned` is `Vec<T>`, so this is basically identical to what it was before.  It just looks shorter because it's easier to access the `insert` method, and the `Vec<T>` is a bit more obfuscated (unfortunately - I actually wanted it to be more obvious, but the only way I could think to do that was to introduce a second type parameter that had no real effect).

File: crates/circuit/src/interner.rs
Comment (jakelishman) [reply crates/circuit/src/interner.rs:33]: Vibes mostly lol.  Logically, `Interned` doesn't _own_ the slice it refers to (e.g. when the `Interned` goes out of scope, the slice doesn't get dropped), it's just a way of retrieving the slice.  That makes it like a reference, except it doesn't have a statically checked lifetime, so `PhantomData<&'a T>` doesn't make sense.

From the perspective of the Rust compiler, there's no difference between me having written `_type: PhantomData<T>` and `_type: PhantomData<*const T>`, I just wrote the latter because that's closer to the actual erased type that `Interned` stores (and in a classical interner, the `Interned` type is often _literally_ a pointer).

Comment (jakelishman) [reply crates/circuit/src/interner.rs:33]: (strictly, I think `Send` would be auto-impl'ed if I'd used `PhantomData<T>` instead of `PhantomData<*const T>`, so maybe there's _technically_ a small difference)

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [reply crates/circuit/src/dag_circuit.rs:1678]: Oh wow, I would not have guessed that. Thanks for clarifying!

I agree with your choice to avoid a redundant type parameter.

File: crates/circuit/src/interner.rs
Comment (kevinhartman) [reply crates/circuit/src/interner.rs:33]: Hmm, alright, that's fine with me. I don't think we'll need to disturb this once it's in place anyhow, so it doesn't seem too important to explain the nuance with a comment etc.

**Comment by kevinhartman** [comment]
Looks good, I'll approve once conflicts are resolved.

File: crates/circuit/src/interner.rs
Comment (jakelishman) [reply crates/circuit/src/interner.rs:33]: My intention had been that this bit of the comment:

> `Interned` is like a non-lifetime-bound reference to data stored in the interner.

was explaining the type of the `PhantomData`.  I'm fine to expand that code comment to be more in depth - I was already trying to answer this question in it anyway.

File: crates/circuit/src/dag_circuit.rs
Comment (jakelishman) [reply crates/circuit/src/dag_circuit.rs]: Done in fb71771c33.

File: crates/circuit/src/interner.rs
Comment (jakelishman) [reply crates/circuit/src/interner.rs:33]: Done in 2d56e17543.

**Review by kevinhartman — APPROVED** [review]

**Pull request merged**
